# Date Fruit Datasets

In [46]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("muratkokludataset/date-fruit-datasets")

print("Path to dataset files:", path)

Path to dataset files: /Users/faisalbalkhair/.cache/kagglehub/datasets/muratkokludataset/date-fruit-datasets/versions/1


In [47]:
import os
os.listdir(path)

['Date_Fruit_Datasets']

In [48]:
inner_path = path + '/Date_Fruit_Datasets'
os.listdir(inner_path)

['Date_Fruit_Datasets_Citation_Request.txt',
 'Date_Fruit_Datasets.xlsx',
 'Date_Fruit_Datasets.arff']

### We use the Date Fruit Dataset to build a ML model that predicts the type of date fruit based on numerical features

In [49]:
import pandas as pd

file_path = inner_path + '/Date_Fruit_Datasets.xlsx'

df = pd.read_excel(file_path)
df.head()

AREA  PERIMETER  MAJOR_AXIS  MINOR_AXIS  ECCENTRICITY   EQDIASQ  \
0  422163   2378.908    837.8484    645.6693        0.6373  733.1539   
1  338136   2085.144    723.8198    595.2073        0.5690  656.1464   
2  526843   2647.394    940.7379    715.3638        0.6494  819.0222   
3  416063   2351.210    827.9804    645.2988        0.6266  727.8378   
4  347562   2160.354    763.9877    582.8359        0.6465  665.2291   

   SOLIDITY  CONVEX_AREA  EXTENT  ASPECT_RATIO  ...  KurtosisRR  KurtosisRG  \
0    0.9947       424428  0.7831        1.2976  ...      3.2370      2.9574   
1    0.9974       339014  0.7795        1.2161  ...      2.6228      2.6350   
2    0.9962       528876  0.7657        1.3150  ...      3.7516      3.8611   
3    0.9948       418255  0.7759        1.2831  ...      5.0401      8.6136   
4    0.9908       350797  0.7569        1.3108  ...      2.7016      2.9761   

   KurtosisRB    EntropyRR    EntropyRG    EntropyRB  ALLdaub4RR  ALLdaub4RG  \
0      4.2287 -59191263232 -50714214400 -39922372608     58.7255     54.9554   
1      3.1704 -34233065472 -37462601728 -31477794816     50.0259     52.8168   
2      4.7192 -93948354560 -74738221056 -60311207936     65.4772     59.2860   
3      8.2618 -32074307584 -32060925952 -29575010304     43.3900     44.1259   
4      4.4146 -39980974080 -35980042240 -25593278464     52.7743     50.9080   

   ALLdaub4RB  Class  
0     47.8400  BERHI  
1     47.8315  BERHI  
2     51.9378  BERHI  
3     41.1882  BERHI  
4     42.6666  BERHI  

[5 rows x 35 columns]

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 898 entries, 0 to 897
Data columns (total 35 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   AREA           898 non-null    int64  
 1   PERIMETER      898 non-null    float64
 2   MAJOR_AXIS     898 non-null    float64
 3   MINOR_AXIS     898 non-null    float64
 4   ECCENTRICITY   898 non-null    float64
 5   EQDIASQ        898 non-null    float64
 6   SOLIDITY       898 non-null    float64
 7   CONVEX_AREA    898 non-null    int64  
 8   EXTENT         898 non-null    float64
 9   ASPECT_RATIO   898 non-null    float64
 10  ROUNDNESS      898 non-null    float64
 11  COMPACTNESS    898 non-null    float64
 12  SHAPEFACTOR_1  898 non-null    float64
 13  SHAPEFACTOR_2  898 non-null    float64
 14  SHAPEFACTOR_3  898 non-null    float64
 15  SHAPEFACTOR_4  898 non-null    float64
 16  MeanRR         898 non-null    float64
 17  MeanRG         898 non-null    float64
 18  MeanRB    

In [51]:
df.columns

Index(['AREA', 'PERIMETER', 'MAJOR_AXIS', 'MINOR_AXIS', 'ECCENTRICITY',
       'EQDIASQ', 'SOLIDITY', 'CONVEX_AREA', 'EXTENT', 'ASPECT_RATIO',
       'ROUNDNESS', 'COMPACTNESS', 'SHAPEFACTOR_1', 'SHAPEFACTOR_2',
       'SHAPEFACTOR_3', 'SHAPEFACTOR_4', 'MeanRR', 'MeanRG', 'MeanRB',
       'StdDevRR', 'StdDevRG', 'StdDevRB', 'SkewRR', 'SkewRG', 'SkewRB',
       'KurtosisRR', 'KurtosisRG', 'KurtosisRB', 'EntropyRR', 'EntropyRG',
       'EntropyRB', 'ALLdaub4RR', 'ALLdaub4RG', 'ALLdaub4RB', 'Class'],
      dtype='object')

In [52]:
df['Class']

0      BERHI
1      BERHI
2      BERHI
3      BERHI
4      BERHI
       ...  
893    SOGAY
894    SOGAY
895    SOGAY
896    SOGAY
897    SOGAY
Name: Class, Length: 898, dtype: object

### We classify the Features and the target

In [53]:
X = df.drop('Class', axis=1)
y = df['Class']

### We split the dataset into 80% Train, 20% Test

In [54]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=42, test_size=0.2, shuffle=True)

### We will train and compare three different models

In [55]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC


logistic_model = LogisticRegression(max_iter=2000)
rForest_model = RandomForestClassifier(random_state=42)
svc_model = SVC()

### We will split the data to 5 parts, 4 to train and 1 to validate

In [56]:
from sklearn.model_selection import cross_val_score

models = {
    'Logistic Regression' : logistic_model,
    'RandomForestClassifier' : rForest_model,
    'SVC' : svc_model

}

for name, model in models.items():
    scores = cross_val_score(model, X_train,y_train, cv=5)
    print(name, scores.mean())

/Users/faisalbalkhair/bootcamp/Week3/Week3_projects/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning:

lbfgs failed to converge after 2000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=2000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/Users/faisalbalkhair/bootcamp/Week3/Week3_projects/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning:

lbfgs failed to converge after 2000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=2000).
You might also want to scale the data as shown in:
    https://scikit-

Logistic Regression 0.6073426573426572
RandomForestClassifier 0.8802544677544677
SVC 0.6365578865578866


In [57]:
rForest_model.fit(X_train,y_train)
score = rForest_model.score(X_test,y_test)
score

0.9111111111111111

### Now I will try to improve the Random Forest Model

In [58]:
random_forset_improved = RandomForestClassifier(
    n_estimators=300,
    random_state=42
)
random_forset_improved.fit(X_train,y_train)

improved_score = random_forset_improved.score(X_test,y_test)
improved_score

0.9111111111111111

### We tried to improve the model by tuning its hyperparameter(such as the number of trees). The final score did not change, thats mean the baseline model was already doing well.

## Count Confusion Matrix:
 
 ! Took it ready from penguins example


In [59]:
from sklearn.metrics import confusion_matrix
import plotly.figure_factory as ff
import numpy as np

def evaluate_model(random_forset_improved, name):
    """Generate and display a confusion matrix heatmap for a given model."""
    y_pred = random_forset_improved.predict(X_test)
    cm = confusion_matrix(y_test, y_pred)
    
    # Create labels for the confusion matrix
    labels = ['DOKOL', 'SAFAVI', 'ROTANA', 'DEGLET', 'SOGAY', 'IRAQI', 'BERHI']
    
    # Create annotated heatmap using plotly
    fig = ff.create_annotated_heatmap(
        z=cm,
        x=labels,
        y=labels,
        colorscale='Blues',
        showscale=True
    )
    
    fig.update_layout(
        title=f'{name} Confusion Matrix',
        xaxis_title='Predicted',
        yaxis_title='Actual',
        width=500,
        height=400
    )
    # Reverse y-axis to match sklearn convention
    fig.update_yaxes(autorange='reversed')
    fig.show()
    
# Evaluate model
evaluate_model(random_forset_improved, "Random Forest")

In [60]:
from sklearn.metrics import accuracy_score, recall_score, precision_score

def get_metrices(model, name):


    y_pred = model.predict(X_test)

    return{
        'Model' : name,
        'Accuracy' : accuracy_score(y_test, y_pred),
        'Precision' : precision_score(y_test, y_pred, average='weighted'),
        'Recall' : recall_score(y_test, y_pred, average='weighted'),
    }

metrics = [
    get_metrices(random_forset_improved, "Random Forest"),
]

# Create a summary DataFrame for easy comparison
df_metrics = pd.DataFrame(metrics)
df_metrics.set_index("Model", inplace=True)
df_metrics.round(4)


Accuracy  Precision  Recall
Model                                     
Random Forest    0.9111      0.924  0.9111

## Save Model

In [61]:
import joblib

joblib.dump(random_forset_improved, "RandomForest_model.pkl")

['RandomForest_model.pkl']

## Final observation:
- The dataset contains 898 samples and 34 features
- Random Forest was better than Logistic regression and svc
- The model achieved 91% on the test data